### PR TITLE
[PSR-15] Provide middleware dispatch examples

### PR DESCRIPTION
**Pull request by weierophinney**
Per discussions on #987, on the mailing list, and in the PSR-15 working group, this patch builds on #987 to add a new section 6.3 to the meta document, detailing "Example Interface Interactions".

The section covers the following:

- A queue-based request handler that aggregates and dispatches middleware.
- A decorator-based request handler for building a layered dispatch mechanism.
- Two examples of re-usable middleware to demonstrate de-coupling from the application and the request handler.
  
This patch incorporates feedback made on #987, but not incorporated there.

### CODE REVIEW

**Comment by weierophinney** [comment]
Ping @Crell — I think the new section answers your questions, but would like for you to review and confirm.

Also pinging @shadowhand — do the changes look okay for you? Also, I rebased off of master, which picked up changes merged since you opened #987.

File: proposed/http-handlers/request-handlers-meta.md [hunk 1]
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
...
+This system has essentially two request handlers, one that will produce a
+response if no middleware generates one, and one for dispatching the middleware
+layers. (In this example, the `RoutingMiddleware` will likely execute composed
+handlers on a successful route match; see more on that below.)
+
+This approach has the following benefits:
+
+- Middleware does not need to know anything about any other middleware or how it
+  is composed in the application.
+- The `QueueRequestHandler` is agnostic of the PSR-7 implementation in use
```
Comment (Crell) [inline proposed/http-handlers/request-handlers-meta.md]: Missing period on this line.

File: proposed/http-handlers/request-handlers-meta.md [hunk 2]
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
...
+$app->add(new RoutingMiddleware());
+
+// execute it:
+$response = $app->handle(ServerRequestFactory::fromGlobals());
+```
+
+This system has essentially two request handlers, one that will produce a
+response if no middleware generates one, and one for dispatching the middleware
+layers. (In this example, the `RoutingMiddleware` will likely execute composed
+handlers on a successful route match; see more on that below.)
```
Comment (Crell) [inline proposed/http-handlers/request-handlers-meta.md]: I don't really see 2 handlers here... I see one handler ($app) that will end up invoked 2-3 times.  Am I missing something or is this worded awkwardly?

File: proposed/http-handlers/request-handlers-meta.md [hunk 3]
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
...
+        $this->nextHandler = $nextHandler;
+    }
+
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        return $this->middleware->process($request, $this->nextHandler);
+    }
+}
+
+$innerHandler = new class ($responsePrototype) implements RequestHandlerInterface {
```
Comment (Crell) [inline proposed/http-handlers/request-handlers-meta.md]: Nitpick: $responsePrototype isn't defined yet.  Is this supposed to be a `new Response())->withStatus(500)`?

File: proposed/http-handlers/request-handlers-meta.md [hunk 4]
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
...
+            ? $corsResult->prepareResponse($this->responsePrototype)
+            : $response;
+    }
+}
+```
+
+Note the use of a "response prototype" in the above example; this approach
+allows the middleware to be agnostic of which PSR-7 implementation is in use,
+and yet still return a response in cases where it determines it should not
+delegate to the handler.Similarly, it is not concerned with how the request
```
Comment (Crell) [inline proposed/http-handlers/request-handlers-meta.md]: Missing space on this line.

File: proposed/http-handlers/request-handlers-meta.md [hunk 5]
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
...
+- Producing a fallback response if no other layer does.
+- Dispatching middleware.
+
+#### Reusable Middleware Examples
+
+In the examples above, we have two middleware composed in each. In order for
+these to work in either situation, we need to write them such that they interact
+appropriately.
+
+Generally speaking, implementors of middleware should follow these guidelines:
```
Comment (Crell) [inline proposed/http-handlers/request-handlers-meta.md]: I feel like these guidelines should be moved to the spec proper, in some form, as they directly inform how middlewares will interoperate.  Eg, when to call the next handler cannot be captured in an interface but is still relevant, and if done "wrong" will lead to a non-interoperable implementation.

File: proposed/http-handlers/request-handlers-meta.md [hunk 6]
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
...
+
+        if ($corsResult->isPreFlight()) {
+            return $corsResult->preparePreFlightResponse($this->responsePrototype);
+        }
+
+        $response = $handler->handle($request);
+
+        return $corsResult->isCorsScoped()
+            ? $corsResult->prepareResponse($this->responsePrototype)
+            : $response;
```
Comment (Crell) [inline proposed/http-handlers/request-handlers-meta.md]: It's cool to see someone besides me that uses/likes that particular formatting idiom. :smile:

File: proposed/http-handlers/request-handlers-meta.md [hunk 7]
```diff
@@ -275,6 +276,10 @@ and a request handler and must return a response. The middleware may:
 - Create and return a response without passing the request to the request handler,
   thereby handling the request itself.
 
+It is expected that a middleware dispatching system will use the request handler
+to delegate response creation to the next available middleware. The last middleware
+would then act as a domain gateway to execute application code.
```
Comment (Crell) [inline proposed/http-handlers/request-handlers-meta.md]: This paragraph... doesn't quite make sense.  I can't wrap my head around what it's saying.  Let me try to rephrase and tell me if I'm in the right ballpark:

```
When delegating from one middleware to the next in a sequence, a dispatching system should use an intermediary request handler as a way to link two middlewares together.  The final or innermost middleware acts as a gateway to application code that will, in most cases, be responsible for response generation.
```

Is that accurate?  (Feel free to steal liberally if so.)  Should that be in the main spec in some form?

**Comment by Crell** [comment]
This is a huge improvement, but still some bits aren't as direct as they could be.  See commentary above.

File: proposed/http-handlers/request-handlers.md [hunk 8]
```diff
@@ -110,20 +110,20 @@ use Psr\Http\Message\ResponseInterface;
...
 
 /**
- * An HTTP middleware component participates in processing an HTTP message,
- * either by acting on the request or the response. This interface defines the
- * methods required to use the middleware.
+ * Participant in processing a server request and response
+ *
+ * An HTTP middleware component participates in processing an HTTP message;
+ * for example, by acting on the request, generating the response, or forwarding the
+ * request to a subsequent middleware and possibly acting on its response.
```
Comment (mnapoli) [inline proposed/http-handlers/request-handlers.md]: 👍

File: proposed/http-handlers/request-handlers-meta.md [hunk 9]
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
 execute the handler provided to the middleware. (This sort of mechanism can even
 be used in conjunction with middleware queues and stacks.)
 
+### 6.3 Example Interface Interactions
+
+The two interfaces, `RequestHandlerInterface` and `MiddlewareInterface`, were
+designed to work in conjunction with one another. While a
+`RequestHandlerInterface` can feasibly be used in isolation, it gains more power
```
Comment (shadowhand) [inline proposed/http-handlers/request-handlers-meta.md]: This sentence isn't necessary, imo. RequestHandler is perfectly viable *without* middleware, not the other way around.

File: proposed/http-handlers/request-handlers-meta.md [hunk 10]
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
...
+- If pre-conditions are met, delegate creation of the response to the provided
+  request handler, optionally providing a "new" request by manipulating the
+  provided request (e.g., `$handler->handle($request->withAttribute('foo',
+  'bar')`).
+
+- Either pass the response returned by the request handler unaltered, or provide
+  a new response by manipulating the one returned (e.g., `return
+  $response->withHeader('X-Foo-Bar', 'baz')`).
+
+The `CorsMiddleware` is one that will exercise all three of these guidelines: if
```
Comment (shadowhand) [inline proposed/http-handlers/request-handlers-meta.md]: I am mildly uncomfortable using a highly specific example like this. This assumes that I know what CORS is and what the failure conditions mean. If we're going to use a specific example perhaps something like routing would be more generally understood?

File: proposed/http-handlers/request-handlers-meta.md [hunk 11]
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
...
+```
+
+Note the use of a "response prototype" in the above example; this approach
+allows the middleware to be agnostic of which PSR-7 implementation is in use,
+and yet still return a response in cases where it determines it should not
+delegate to the handler.Similarly, it is not concerned with how the request
+handler is implemented; it merely uses it to produce a response when
+pre-conditions have been met.
+
+The `RoutingMiddleware` implementation described below follows a similar
```
Comment (shadowhand) [inline proposed/http-handlers/request-handlers-meta.md]: Yes this is better, but I think the process should be:

- If routing matches, continue processing.
- Else return a `HTTP Not Found` response.

File: proposed/http-handlers/request-handlers.md [hunk 12]
```diff
@@ -112,8 +112,8 @@ use Psr\Http\Message\ServerRequestInterface;
 /**
  * Participant in processing a server request and response
  *
- * An HTTP middleware component participates in processing an HTTP message,
- * either by acting on the request, generating the response, or forwarding the
+ * An HTTP middleware component participates in processing an HTTP message;
+ * for example, by acting on the request, generating the response, or forwarding the
```
Comment (shadowhand) [inline proposed/http-handlers/request-handlers.md]: Why the change of phrasing? "Either" seems like the appropriate wording.

File: proposed/http-handlers/request-handlers-meta.md (hunk 2 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: No, you're correct. The original example I created composed another request handler instead of a request prototype, and delegated to that handler if the last middleware in the queue calls on the handler instead of returning a response. I'm going to refactor the example to to do that, as it also then answers the "request prototype" question you have elsewhere.

File: proposed/http-handlers/request-handlers-meta.md (hunk 5 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: @shadowhand What do you think about moving these into the spec proper? If we do, where should they live?

File: proposed/http-handlers/request-handlers-meta.md (hunk 3 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: It could be, or a 404 status. I'll update to make this more clear.

File: proposed/http-handlers/request-handlers-meta.md (hunk 10 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: I'll see if I can come up with another example. This was one I liked because it hits all the various ways middleware might produce a response:

- on its own, due to a failed pre-condition
- by acting as a pass-through to the request handler
- by modifying the response returned by the request handler

Let me see if I can come up with an example, and I'll ping you if/when I do.

File: proposed/http-handlers/request-handlers-meta.md (hunk 11 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: That's actually exactly how we do it in Expressive. :+1:

The only issue I see with that is that it complicates the example: I then need to add dispatch middleware and a response prototype or factory. I felt the "map route to handler" approach was sufficiently simple and didn't require extra explanation.

File: proposed/http-handlers/request-handlers.md (hunk 12 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers.md]: This was based on feedback from @mnapoli, IIRC.

File: proposed/http-handlers/request-handlers-meta.md (hunk 10 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: Came up with an example using authorization instead of CORS; please review and let me know what you think.

File: proposed/http-handlers/request-handlers-meta.md (hunk 7 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: It's correct, except that the "application code that will, in most cases, be responsible for response generation" will likely either be the middleware itself, or the a _request handler_. I'll see if I can come up with a re-phrase that makes sense.

File: proposed/http-handlers/request-handlers-meta.md [hunk 13]
```diff
@@ -276,9 +276,11 @@ and a request handler and must return a response. The middleware may:
...
   thereby handling the request itself.
 
-It is expected that a middleware dispatching system will use the request handler
-to delegate response creation to the next available middleware. The last middleware
-would then act as a domain gateway to execute application code.
+When delegating from one middleware to the next in a sequence, a dispatching
+system should use an intermediary request handler as a way to link middleware
+together. The final or innermost middleware will act as a gateway to application
+code and generate a response from its results; alternately, the middleware may
+delegate this responsibility to a _request handler_.
```
Comment (weierophinney) [inline proposed/http-handlers/request-handlers-meta.md]: @Crell This is the updated text, using your suggestion as a starting point.

File: proposed/http-handlers/request-handlers-meta.md [hunk 14]
```diff
@@ -368,6 +370,240 @@ executing that handler. If unable to route to a handler, it would instead
...
+
+- If authorization is required, but the request is not authorized, it will use a
+  composed prototype response to produce an "unauthorized" response.
+- If authorization is not required, it will delegate the request to the handler
+  without changes.
+- If authorization is required and the request is authorized, it will delegate
+  the request to the handler, and sign the response returned based on the request.
+
+```php
+class AuthorizationMiddleware implements MiddlewareInterface
```
Comment (weierophinney) [inline proposed/http-handlers/request-handlers-meta.md]: @shadowhand Let me know if this altered example (replacing CorsMiddleware with AuthorizationMiddleware) makes more sense. I definitely want to demonstrate multiple middleware, and particularly different approaches middleware can take both to be re-usable as well as to change the workflow/paths of an application.

File: proposed/http-handlers/request-handlers-meta.md [hunk 15]
```diff
@@ -368,6 +370,240 @@ executing that handler. If unable to route to a handler, it would instead
...
+        AuthorizationMap $authorizationMap,
+        ResponseInterface $responsePrototype
+    ) {
+        $this->authorizationMap = $authorizationMap;
+        $this->responsePrototype = $responsePrototype;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler) : ResponseInterface
+    {
+        if (! $authorizationMap->requestRequiresAuthorization($request)) {
```
Comment (shadowhand) [inline proposed/http-handlers/request-handlers-meta.md]: How about `doesRequireAuthorization` instead of `requestRequiresAuthorization`?

File: proposed/http-handlers/request-handlers-meta.md [hunk 16]
```diff
@@ -368,6 +370,240 @@ executing that handler. If unable to route to a handler, it would instead
...
+
+    public function __construct(
+        AuthorizationMap $authorizationMap,
+        ResponseInterface $responsePrototype
+    ) {
+        $this->authorizationMap = $authorizationMap;
+        $this->responsePrototype = $responsePrototype;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler) : ResponseInterface
```
Comment (shadowhand) [inline proposed/http-handlers/request-handlers-meta.md]: `): ResponseInterface` as per purposed PSR-12 spec, no space.

https://github.com/php-fig/fig-standards/blob/master/proposed/extended-coding-style-guide.md#45-method-and-function-arguments

File: proposed/http-handlers/request-handlers-meta.md [hunk 17]
```diff
@@ -368,6 +370,240 @@ executing that handler. If unable to route to a handler, it would instead
...
+{
+    private $authorizationMap;
+    private $responsePrototype;
+
+    public function __construct(
+        AuthorizationMap $authorizationMap,
+        ResponseInterface $responsePrototype
+    ) {
+        $this->authorizationMap = $authorizationMap;
+        $this->responsePrototype = $responsePrototype;
```
Comment (shadowhand) [inline proposed/http-handlers/request-handlers-meta.md]: I'm not sure why we need a response prototype here. I think it would make more sense of the `AuthorizationMap` was responsible for generating/manipulating the response. That would simplify the example and keep the focus on the middleware.

File: proposed/http-handlers/request-handlers-meta.md (hunk 5 above)
Comment (shadowhand) [reply proposed/http-handlers/request-handlers-meta.md]: I don't think we should but putting code samples in the spec itself. Having the guidelines at the top of the spec would be good though.

File: proposed/http-handlers/request-handlers-meta.md (hunk 15 above)
Comment (Jean85) [reply proposed/http-handlers/request-handlers-meta.md]: Maybe `needsAuthorization` to keep it short? (for easiness of readability)

File: proposed/http-handlers/request-handlers-meta.md (hunk 5 above)
Comment (Crell) [reply proposed/http-handlers/request-handlers-meta.md]: Yes, I was referring to the guidelines, (probably written a bit differently).  I agree the code samples should remain in the metadoc.

File: proposed/http-handlers/request-handlers.md (hunk 12 above)
Comment (Crell) [reply proposed/http-handlers/request-handlers.md]: This seems fine.  If we want to get super pedantic then "either" is technically only correct when there's 2 options, and in this case there are 3.  "For example" seems OK to me.

File: proposed/http-handlers/request-handlers-meta.md (hunk 13 above)
Comment (Crell) [reply proposed/http-handlers/request-handlers-meta.md]: Much clearer!  Why is the last "request handler" set off, though?  Perhaps "to a dedicated request handler"?  (Nitpicking here.)

A statement of this sort seems like it belongs in the spec.  It's not a "why and background", it's a "no, really, this is how you're supposed to use it".  That seems like it should be in the spec proper.

File: proposed/http-handlers/request-handlers.md (hunk 12 above)
Comment (Crell) [reply proposed/http-handlers/request-handlers.md]: Or, technically, with the semicolon we could omit that word entirely and it would still be a grammatically valid list of the available options.

File: proposed/http-handlers/request-handlers-meta.md [hunk 18]
```diff
@@ -368,6 +370,240 @@ executing that handler. If unable to route to a handler, it would instead
...
+```php
+class AuthorizationMiddleware implements MiddlewareInterface
+{
+    private $authorizationMap;
+    private $responsePrototype;
+
+    public function __construct(
+        AuthorizationMap $authorizationMap,
+        ResponseInterface $responsePrototype
+    ) {
```
Comment (Crell) [inline proposed/http-handlers/request-handlers-meta.md]: There's only 2 parameters. No need to go with the harder-to-read multi-line approach.  Icky.

File: proposed/http-handlers/request-handlers-meta.md (hunk 5 above)
Comment (mnapoli) [reply proposed/http-handlers/request-handlers-meta.md]: I really think the spec should not dictate *how* middlewares should behave (or other implementation details), e.g. when to generate a response or call the delegate.

Maybe that's not what you meant, but the guidelines below are too precise IMO.

Comment (stefanotorresi) [reply proposed/http-handlers/request-handlers-meta.md]: I agree with @shadowhand and @mnapoli: I don't think the spec document should contain these guidelines.

Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: @mnapoli and @stefanotorresi — thanks; I think that's great feedback, and I tend to agree. The guidelines presented are implementation details and good practices, but _do not need to be followed in order to comply with the specification_. I'll leave them where they are.

File: proposed/http-handlers/request-handlers-meta.md (hunk 18 above)
Comment (shadowhand) [reply proposed/http-handlers/request-handlers-meta.md]: When using objects as type, I prefer the multi-line approach. Easier to read and less churn on diffs when parameters change.

File: proposed/http-handlers/request-handlers-meta.md (hunk 17 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: Agreed; making that change now.

File: proposed/http-handlers/request-handlers-meta.md (hunk 18 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: I was going to disagree, but when I re-formatted, I realized the line is still shorter than the `process()` method declaration line. About to commit that now.

Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: Plus, with the change to remove the response prototype, it's now only one argument anyways...

**Comment by weierophinney** [comment]
Requested changes have been made (mostly):

- I **did not** move the "guidelines" bullet points to the spec; as several folks noted, coding to those guidelines will definitely mean you're coding to the spec, but you do not need to follow them to implement the spec.
- I removed the emphasis around "request handler", and added the word "dedicated" before them, per @Crell.
- I moved that particular paragraph to section 1.2 of the spec, per @Crell.
- I updated the preamble to `MiddlewareInterface` to incorporate the various bits of feedback.
- I updated code samples to follow PSR-12.
- I made changes to the `AuthorizationMiddleware` example per feedback received.

@Crell, @shadowhand, @mnapoli and @stefanotorresi — can you give one more review so we can finalize these? I think we can then prepare for a CC vote.

File: proposed/http-handlers/request-handlers-meta.md (hunk 5 above)
Comment (mnapoli) [reply proposed/http-handlers/request-handlers-meta.md]: Could you make that part clear in the text? Because it reads like a specification (with the "should" keyword).

Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: I'm not sure what other word to user here, TBH, and the fact that it's not ALL CAPS means it does not have the same semantic weight per RFC 2119. Maybe the following?

> Implementors of middleware striving for maximum interoperability may want to consider the following guidelines:

Would that work for you, @mnapoli ?

File: proposed/http-handlers/request-handlers.md [hunk 19]
```diff
@@ -52,6 +52,12 @@ the creation of a resulting response, as defined by PSR-7.
 A middleware component MAY create and return a response without delegating to
 a request handler, if sufficient conditions are met.
 
+When delegating from one middleware to the next in a sequence, a dispatching
+system should use an intermediary request handler as a way to link middleware
+together. The final or innermost middleware will act as a gateway to application
+code and generate a response from its results; alternately, the middleware may
+delegate this responsibility to a dedicated request handler.
```
Comment (Crell) [inline proposed/http-handlers/request-handlers.md]: Nitpick: SHOULD and MAY should be capitalized to mean the defined SHOULD and MAY, I think, since this is in the spec proper.

Comment (weierophinney) [reply proposed/http-handlers/request-handlers.md]: My understanding is they should only be capitalized when they have the semantic meanings as defined in RFC 2119; the question is whether or not those words here have those semantic meanings. I did not originally write this with those meanings in mind.

The problem, of course, is that there are very few ways to re-word this to have the same intended meaning _without_ using those words. Can you propose an alternative? Or do you feel they have those semantics?

File: proposed/http-handlers/request-handlers-meta.md [hunk 20]
```diff
@@ -368,6 +364,233 @@ executing that handler. If unable to route to a handler, it would instead
...
+  "404 Not Found" condition, a default page, etc.
+
+#### Decoration-based request handler
+
+In this approach, a request handler implementation decorates both a middleware
+instance and a fallback request handler to pass to it. The application is built
+from the outside-in, passing each request handler "layer" to the next outer one.
+
+```php
+class DecoratedRequestHandler implements RequestHandlerInterface
```
Comment (Crell) [inline proposed/http-handlers/request-handlers-meta.md]: Nitpick: Technically, this isn't a decoratED handler.  It's a decoratING handler.  The name should be more descriptive.  Sorry for not catching that before.

File: proposed/http-handlers/request-handlers.md (hunk 19 above)
Comment (ashnazg) [reply proposed/http-handlers/request-handlers.md]: If not RFC2119, maybe `can` could be used for both?

File: proposed/http-handlers/request-handlers-meta.md (hunk 20 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: Fixed.

File: proposed/http-handlers/request-handlers.md (hunk 19 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers.md]: `CAN` also has semantic meaning via RFC 2119. :)

Been discussing this and the idea of moving the section to the spec document itself heavily today; hopefully we can have resolution shortly.

Comment (ashnazg) [reply proposed/http-handlers/request-handlers.md]: Hmm, I don't see `CAN` mentioned in the RFC... hence me thinking it safe.

Would something wordy, like `could potentially`, do the job without sounding like a 2119 directive?

**Comment by weierophinney** [comment]
At this time, based on discussion with the working group, I've moved the paragraph in question back to the meta document's section 6.2 preamble, with changes as discussed here and elsewhere to remove implications that these are anything more than optional recommendations.

File: proposed/http-handlers/request-handlers-meta.md (hunk 5 above)
Comment (weierophinney) [reply proposed/http-handlers/request-handlers-meta.md]: Updated!
